dist/tools/tunslip: Fix some ugly code

gebart opened the pull request:
The tunslip programs were imported from Contiki, and they are probably among the ugliest pieces of code in both Contiki and RIOT. This PR fixes some apparent issues in tunslip6 and makes the warnings visible for tapslip6  and tunslip, to let the user know that it is ugly code.

Added a workaround for a bug where an IPv6 header containing `'\n'` (0x0a) was mistaken for a printable string. <- This was the reason I ventured into this dark corner of the Contiki repo originally.

Also added a `clean` target to dist/tools/tunslip/Makefile 

Tested tunslip6 against Contiki's `examples/ipv6/rpl-border-router` border router implementation.

miri64 commented:
Change looks alright, but out of interest: Did you also propose upstream? Can only test after Monday, since I don't have a UART adapter at hand.

gebart commented:
I did not yet provide an upstream PR. I'll post here when I have a PR number

lebrush commented on dist/tools/tunslip/tunslip6.c:
```diff
@@ -147,18 +158,34 @@ stamptime(void)
         else if (s[i] < ' ' || '~' < s[i]) {
             return 0;
         }
+        /* only return 1 if the string contains at least one letter or number */
```
Why only [A-Za-z0-9] are allowed to be printed?

gebart replied:
It's not, it's requiring at least one letter or number before it can accept the given string as printable. It's there to avoid misinterpreting parts of packets as "strings" such as mistaking
```
"`'  \r \0 "^~\n"
```
for a valid string and printing it instead of leaving it in the packet buffer.
I reasoned that by requiring at least one letter or number in a string then we at least reduce the risk somewhat.

lebrush replied:
Thanks for the clarification it definitely make sense :-)

Just an idea, what about just dumping all packets as an hexdump (see `hexdump -C`) so all of them are printed in the same way?

Example:
```
40 8b 88 48 5a cc 9a 35  8b 65 cb 96 75 f5 6e 74  |@..HZ..5.e..u.nt|
68 c1 82 05 3c f0 c0 03  5d bd 1b 22 22 22 22 49  |h...<...]..""""I|
```

(this relates as well to your comment: https://github.com/RIOT-OS/RIOT/pull/5677#discussion_r71969288)

gebart replied:
The reason that the tunslip application tries to find printable strings is to be able to multiplex debug messages with SLIP traffic. If you want a packet dump you can run tunslip with the switch `-v4` and everything will be dumped to the terminal as well as passed on to the tun interface.

gebart commented on dist/tools/tunslip/tunslip6.c:
```diff
@@ -147,18 +158,34 @@ stamptime(void)
 
-    for (i = 1; i < len; i++) {
+    int ret = 0;
+    for (int i = 1; i < len; i++) {
```
I wonder if this is not off by one? I think `i` should start at 0..

miri64 replied:
If you can fix it and its working for you, I say go for it ;-)

gebart commented on dist/tools/tunslip/tunslip6.c:
```diff
@@ -147,18 +158,34 @@ stamptime(void)
-    for (i = 1; i < len; i++) {
+    int ret = 0;
+    for (int i = 1; i < len; i++) {
         if (s[i] == 0 || s[i] == '\r' || s[i] == '\n' || s[i] == '\t') {
```
Does it make sense to interpret null chars as printable?

miri64 replied:
From my earlier experiences with SLIP and terminal multiplexing implementations: `\0` is just a good signal character to point out that something is printable: IPv4 packets start with `0x4` half-byte and IPv6 start with `0x6` half-byte. But I guess there are other reasons `\0` is included here.

miri64 commented:
Changes look okay code-wise, but I have no means to test it at the moment. @kYc0o would you be willing to take this one over?

miri64 commented:
Ping @kYc0o?

kYc0o commented:
I can give a try, let's see during next week.

kYc0o commented:
I think clang is very pedantic...

```console
cc -Wall -Wextra -pedantic -std=c99    tapslip6.c   -o tapslip6
tapslip6.c:167:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tapslip6.c:267:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tapslip6.c:269:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tapslip6.c:313:12: warning: unused variable 'ecode' [-Wunused-variable]
    int i, ecode;
           ^
tapslip6.c:368:18: error: use of undeclared identifier 'B115200'
speed_t b_rate = BAUDRATE;
                 ^
tapslip6.c:366:18: note: expanded from macro 'BAUDRATE'
#define BAUDRATE B115200
                 ^
tapslip6.c:385:5: warning: implicit declaration of function 'cfmakeraw' is invalid in C99 [-Wimplicit-function-declaration]
    cfmakeraw(&tty);
    ^
tapslip6.c:390:21: error: use of undeclared identifier 'CRTSCTS'
    tty.c_cflag &= ~CRTSCTS;
                    ^
tapslip6.c:507:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tapslip6.c:559:5: error: use of undeclared identifier 'fd_set'
    fd_set rset, wset;
    ^
tapslip6.c:629:22: error: use of undeclared identifier 'B57600'
            b_rate = B57600;
                     ^
tapslip6.c:633:22: error: use of undeclared identifier 'B115200'
            b_rate = B115200;
                     ^
tapslip6.c:696:9: warning: implicit declaration of function 'FD_ZERO' is invalid in C99 [-Wimplicit-function-declaration]
        FD_ZERO(&rset);
        ^
tapslip6.c:696:18: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        FD_ZERO(&rset);
                 ^~~~
                 ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
tapslip6.c:697:18: error: use of undeclared identifier 'wset'
        FD_ZERO(&wset);
                 ^
tapslip6.c:719:13: warning: implicit declaration of function 'FD_SET' is invalid in C99 [-Wimplicit-function-declaration]
            FD_SET(slipfd, &wset);
            ^
tapslip6.c:719:29: error: use of undeclared identifier 'wset'
            FD_SET(slipfd, &wset);
                            ^
tapslip6.c:722:25: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        FD_SET(slipfd, &rset);  /* Read from slip ASAP! */
                        ^~~~
                        ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
tapslip6.c:730:28: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            FD_SET(tunfd, &rset);
                           ^~~~
                           ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
tapslip6.c:737:15: warning: implicit declaration of function 'select' is invalid in C99 [-Wimplicit-function-declaration]
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
              ^
tapslip6.c:737:34: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
                                 ^~~~
                                 ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
tapslip6.c:737:41: error: use of undeclared identifier 'wset'
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
                                        ^
tapslip6.c:743:17: warning: implicit declaration of function 'FD_ISSET' is invalid in C99 [-Wimplicit-function-declaration]
            if (FD_ISSET(slipfd, &rset)) {
                ^
tapslip6.c:743:35: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            if (FD_ISSET(slipfd, &rset)) {
                                  ^~~~
                                  ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
tapslip6.c:747:35: error: use of undeclared identifier 'wset'
            if (FD_ISSET(slipfd, &wset)) {
                                  ^
tapslip6.c:752:50: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            if (slip_empty() && FD_ISSET(tunfd, &rset)) {
                                                 ^~~~
                                                 ret
tapslip6.c:558:9: note: 'ret' declared here
    int ret;
        ^
10 warnings and 15 errors generated.
```

this is for `tapslip6`, where I don't see a lot of changes, but actually it is very ugly too.

for `tunslip6`

```console
cc -Wall -Wextra -pedantic -std=c99    tunslip6.c   -o tunslip6
tunslip6.c:590:18: error: use of undeclared identifier 'B115200'
speed_t b_rate = BAUDRATE;
                 ^
tunslip6.c:588:18: note: expanded from macro 'BAUDRATE'
#define BAUDRATE B115200
                 ^
tunslip6.c:607:5: warning: implicit declaration of function 'cfmakeraw' is invalid in C99 [-Wimplicit-function-declaration]
    cfmakeraw(&tty);
    ^
tunslip6.c:614:24: error: use of undeclared identifier 'CRTSCTS'
        tty.c_cflag |= CRTSCTS;
                       ^
tunslip6.c:617:25: error: use of undeclared identifier 'CRTSCTS'
        tty.c_cflag &= ~CRTSCTS;
                        ^
tunslip6.c:735:26: warning: implicit declaration of function 'index' is invalid in C99 [-Wimplicit-function-declaration]
        char   *prefix = index(itfaddr, '/');
                         ^
tunslip6.c:735:17: warning: incompatible integer to pointer conversion initializing 'char *' with an expression of type 'int' [-Wint-conversion]
        char   *prefix = index(itfaddr, '/');
                ^        ~~~~~~~~~~~~~~~~~~~
tunslip6.c:877:15: warning: incompatible integer to pointer conversion initializing 'char *' with an expression of type 'int' [-Wint-conversion]
        char *prefix = index(itfaddr, '/');
              ^        ~~~~~~~~~~~~~~~~~~~
tunslip6.c:1072:22: error: use of undeclared identifier 'B57600'
            b_rate = B57600;
                     ^
tunslip6.c:1076:22: error: use of undeclared identifier 'B115200'
            b_rate = B115200;
                     ^
tunslip6.c:1080:22: error: use of undeclared identifier 'B230400'
            b_rate = B230400;
                     ^
4 warnings and 6 errors generated.
```

And finally `tunslip`

```console
cc -Wall -Wextra -pedantic -std=c99    tunslip.c   -o tunslip
tunslip.c:317:9: warning: expression result unused [-Wunused-value]
        *p++;
        ^~~~
tunslip.c:561:59: warning: passing 'unsigned char [2000]' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                if (ecode < 0 && inbufptr == 8 && strncmp(uip.inbuf, "=IPA", 4) == 0) {
                                                          ^~~~~~~~~
/usr/include/string.h:84:26: note: passing argument to parameter here
int      strncmp(const char *, const char *, size_t);
                             ^
tunslip.c:530:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tunslip.c:667:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tunslip.c:669:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tunslip.c:795:18: error: use of undeclared identifier 'B115200'
speed_t b_rate = BAUDRATE;
                 ^
tunslip.c:793:18: note: expanded from macro 'BAUDRATE'
#define BAUDRATE B115200
                 ^
tunslip.c:812:5: warning: implicit declaration of function 'cfmakeraw' is invalid in C99 [-Wimplicit-function-declaration]
    cfmakeraw(&tty);
    ^
tunslip.c:817:21: error: use of undeclared identifier 'CRTSCTS'
    tty.c_cflag &= ~CRTSCTS;
                    ^
tunslip.c:933:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tunslip.c:985:5: error: use of undeclared identifier 'fd_set'
    fd_set rset, wset;
    ^
tunslip.c:1063:22: error: use of undeclared identifier 'B57600'
            b_rate = B57600;
                     ^
tunslip.c:1067:22: error: use of undeclared identifier 'B115200'
            b_rate = B115200;
                     ^
tunslip.c:1211:9: warning: implicit declaration of function 'FD_ZERO' is invalid in C99 [-Wimplicit-function-declaration]
        FD_ZERO(&rset);
        ^
tunslip.c:1211:18: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        FD_ZERO(&rset);
                 ^~~~
                 ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1212:18: error: use of undeclared identifier 'wset'
        FD_ZERO(&wset);
                 ^
tunslip.c:1225:13: warning: implicit declaration of function 'FD_SET' is invalid in C99 [-Wimplicit-function-declaration]
            FD_SET(slipfd, &wset);
            ^
tunslip.c:1225:29: error: use of undeclared identifier 'wset'
            FD_SET(slipfd, &wset);
                            ^
tunslip.c:1228:25: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        FD_SET(slipfd, &rset);  /* Read from slip ASAP! */
                        ^~~~
                        ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1236:28: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            FD_SET(tunfd, &rset);
                           ^~~~
                           ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1243:33: error: use of undeclared identifier 'rset'; did you mean 'ret'?
                FD_SET(dhsock, &rset);
                                ^~~~
                                ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1251:15: warning: implicit declaration of function 'select' is invalid in C99 [-Wimplicit-function-declaration]
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
              ^
tunslip.c:1251:34: error: use of undeclared identifier 'rset'; did you mean 'ret'?
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
                                 ^~~~
                                 ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1251:41: error: use of undeclared identifier 'wset'
        ret = select(maxfd + 1, &rset, &wset, NULL, NULL);
                                        ^
tunslip.c:1257:17: warning: implicit declaration of function 'FD_ISSET' is invalid in C99 [-Wimplicit-function-declaration]
            if (FD_ISSET(slipfd, &rset)) {
                ^
tunslip.c:1257:35: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            if (FD_ISSET(slipfd, &rset)) {
                                  ^~~~
                                  ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1261:35: error: use of undeclared identifier 'wset'
            if (FD_ISSET(slipfd, &wset)) {
                                  ^
tunslip.c:1266:50: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            if (slip_empty() && FD_ISSET(tunfd, &rset)) {
                                                 ^~~~
                                                 ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1272:67: error: use of undeclared identifier 'rset'; did you mean 'ret'?
            if (dhsock != -1 && slip_empty() && FD_ISSET(dhsock, &rset)) {
                                                                  ^~~~
                                                                  ret
tunslip.c:984:9: note: 'ret' declared here
    int ret;
        ^
tunslip.c:1096:15: warning: comparison of integers of different signs: 'in_addr_t' (aka 'unsigned int') and 'int' [-Wsign-compare]
        if (a == -1) {
            ~ ^  ~~
12 warnings and 17 errors generated.
```

It's weird but `tunslip6` from contiki's upstream works fine in my Mac, I made a diff and there's no differences which could break it on RIOT.

Besides the error about the baudrates (contiki define them in a `tools-utils.h`), the problem about `CRTSCTS` (included in `termios.h`) it's maybe related to the compilation flags (maybe `-std=c99`).

kYc0o commented:
Apparently this breaks heavily clang on OS X:

```c
/* need POSIX functions */
#if __STDC_VERSION__ >= 199901L
#define _XOPEN_SOURCE 600
#else
#define _XOPEN_SOURCE 500
#endif /* __STDC_VERSION__ */
```

Without this `tunslip6` compiles correctly (even without `tools-utils`) and the only warnings I have are about bad coding:

```console
cc -Wall -Wextra -pedantic -std=c99    tapslip6.c   -o tapslip6
tapslip6.c:167:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tapslip6.c:267:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tapslip6.c:269:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tapslip6.c:313:12: warning: unused variable 'ecode' [-Wunused-variable]
    int i, ecode;
           ^
tapslip6.c:507:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tapslip6.c:102:1: warning: unused function 'print_packet' [-Wunused-function]
print_packet(u_int8_t *p, int len)
^
6 warnings generated.
cc -Wall -Wextra -pedantic -std=c99    tunslip.c   -o tunslip
tunslip.c:317:9: warning: expression result unused [-Wunused-value]
        *p++;
        ^~~~
tunslip.c:561:59: warning: passing 'unsigned char [2000]' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                if (ecode < 0 && inbufptr == 8 && strncmp(uip.inbuf, "=IPA", 4) == 0) {
                                                          ^~~~~~~~~
/usr/include/string.h:84:26: note: passing argument to parameter here
int      strncmp(const char *, const char *, size_t);
                             ^
tunslip.c:530:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tunslip.c:667:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tunslip.c:669:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tunslip.c:933:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tunslip.c:1096:15: warning: comparison of integers of different signs: 'in_addr_t' (aka 'unsigned int') and 'int' [-Wsign-compare]
        if (a == -1) {
            ~ ^  ~~
7 warnings generated.
cc -Wall -Wextra -pedantic -std=c99    tunslip6.c   -o tunslip6
```

gebart commented:
@kYc0o the Contiki Makefile simply builds it without any extra flags `cc tunslip6.c tools-utils.c   -o tunslip6`, which is why the warnings don't show up in the Contiki build.

gebart commented:
added 3 more commits which were lying around and rebased.
I'm unsure whether https://github.com/RIOT-OS/RIOT/pull/5677/commits/c41a4ac41b404beb4150628c6bf901ff950cec7f should stay or go

kYc0o commented:
Yeah I know, but after removing the code in [this](https://github.com/RIOT-OS/RIOT/pull/5677#issuecomment-253243368) comment tunslip6 compiles without any warnings.

kYc0o commented:
Then what's the final state of this? We drop (again) OS X support for this tool? (requires rebase btw).

miri64 commented:
@kYc0o is anyone using SLIP at the moment?

kYc0o commented:
I don't know but for me it seems useful since I didn't have time to adapt ethos on OS X...

miri64 commented:
So there is still merit for me to fix SLIP :astonished:

kYc0o commented:
Ping!

gebart commented:
rebased and removed the contested _XOPEN_SOURCE stuff. It seems to build fine on my system with both GCC and Clang now.

Tool versions used:
gcc (Gentoo 5.4.0 p1.0, pie-0.6.5) 5.4.0
clang version 3.9.0 (tags/RELEASE_390/final)
glibc (Gentoo 2.23-r3 p5) 2.23

kYc0o commented:
Ok great! Now tunslip6 compiles without any issue. However `tunslip` and `tapslip6` still have problems (with  _XOPEN_SOURCE stuff removed):

```console
cc -Wall -Wextra -pedantic -std=c99    tunslip.c   -o tunslip
tunslip.c:317:9: warning: expression result unused [-Wunused-value]
        *p++;
        ^~~~
tunslip.c:561:59: warning: passing 'unsigned char [2000]' to parameter of type 'const char *' converts between pointers to integer types with different sign
      [-Wpointer-sign]
                if (ecode < 0 && inbufptr == 8 && strncmp(uip.inbuf, "=IPA", 4) == 0) {
                                                          ^~~~~~~~~
/usr/include/string.h:84:26: note: passing argument to parameter here
int      strncmp(const char *, const char *, size_t);
                             ^
tunslip.c:530:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tunslip.c:667:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tunslip.c:669:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tunslip.c:933:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tunslip.c:1095:15: warning: comparison of integers of different signs: 'in_addr_t' (aka 'unsigned int') and 'int' [-Wsign-compare]
        if (a == -1) {
            ~ ^  ~~
7 warnings generated.
```

And:

```console
cc -Wall -Wextra -pedantic -std=c99    tapslip6.c   -o tapslip6
tapslip6.c:167:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (inbufptr >= sizeof(uip.inbuf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~
tapslip6.c:267:15: warning: unused parameter 'fd' [-Wunused-parameter]
slip_send(int fd, unsigned char c)
              ^
tapslip6.c:269:18: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    if (slip_end >= sizeof(slip_buf)) {
        ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
tapslip6.c:507:14: warning: unused parameter 'signo' [-Wunused-parameter]
sigalarm(int signo)
             ^
tapslip6.c:102:1: warning: unused function 'print_packet' [-Wunused-function]
print_packet(u_int8_t *p, int len)
^
5 warnings generated.
```

I'm on clang:
```
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
```

gebart commented:
I'll remove the XOPEN defines from tunslip and tapslip. This PR is not intended to fix all problems with these programs, only make them a bit more compliant. I don't think tunslip sees much use in RIOT because most development is focused around IPv6.

gebart commented:
@kYc0o 
ping

kYc0o commented:
Please squash to give Murdock a chance.

kYc0o commented:
Restarting Murdock since it caught the label "NEEDS SQUASHING"

kYc0o commented:
All green -> Go!
